tuf-spec.md

```diff
@@ -736,18 +735,23 @@ repo](https://github.com/theupdateframework/specification/issues).
          , ...
        }
 
-   METAPATH is the metadata file's path on the repository relative to the
-   metadata base URL.
+   METAPATH is the file path of the metadata on the repository relative to the
+   metadata base URL. For snapshot.json, these are top-level targets metadata
+   and delegated targets metadata.
 
-   VERSION is listed for the top-level targets and all delegated targets roles
-   available on the repository.
+   VERSION is the integer version number as shown in the metadata file at
+   METAPATH.
 
-   LENGTH is the integer length in bytes of the metadata file. It is
-   OPTIONAL for all roles.
+   LENGTH is the integer length in bytes of the metadata file at METAPATH. It
+   is OPTIONAL and can be omitted to reduce the snapshot metadata file size. In
+   that case the client MUST use a custom download limit for the listed
+   metadata.
 
-   HASHES is the dictionary that specifies one or more hashes, including
-   the cryptographic hash function.  For example: { "sha256": HASH, ... }. It is
-   OPTIONAL for all roles.
+   HASHES is a dictionary that specifies one or more hashes of the metadata
+   file at METAPATH, including their cryptographic hash function. For example:
+   { "sha256": HASH, ... }. HASHES is OPTIONAL and can be omitted to reduce
+   the snapshot metadata file size.  In that case the repository MUST guarantee
+   that VERSION alone unambiguously identifies the metadata at METAPATH.
```

Comment:
We might want to note when length and hashes should be included (ie backwards compatibility for both, using length as an exact download size)

Reply:
Backwards compatibility is one reason, yes, but I believe the more important reason is to protect against certain security attacks. Please see Section 5.6 of [Mercury](https://www.usenix.org/system/files/conference/atc17/atc17-kuppusamy.pdf).

Reply:
Also, if we are bringing back hashes and length here, we should add the corresponding checks in the download process.

Reply:
>We might want to note when length and hashes should be included

@mnm678, @trishankatdatadog: Good idea. Could you push a commit with such a note?

> Also, if we are bringing back hashes and length here, we should add the corresponding checks in the download process.

@trishankatdatadog: What's missing from the existing ([L1136-1138](https://github.com/theupdateframework/specification/pull/90/files#diff-94b2d37a08c587082c1ea1aaead6f973R1136-R1138)) and newly proposed ([L11160-1163](https://github.com/theupdateframework/specification/pull/90/files#diff-94b2d37a08c587082c1ea1aaead6f973R1160-R1163)) download checks?

Reply:
@lukpueh Those are good checks, but:

* 3.1 doesn't mention that hashes and length are optional. It also doesn't mention checking the length, if any, of the `snapshot` metadata.
* When we check the `targets` role or its delegations, do we mention checking optional hashes and length from the `snapshot` metadata?

Reply:
Thanks for pointing these out, @trishankatdatadog!

> * 3.1 doesn't mention that hashes and length are optional. It also doesn't mention checking the length, if any, of the `snapshot` metadata.

Is another length match check in 3.1. necessary? We already say in 3. that the client should download snapshot.json up to the length specified in timestamp.json (if specified).

Regarding optionality of hashes, I agree we could add a "(if any)" note in 3.1, as we do for targets in 4.1.

> * When we check the `targets` role or its delegations, do we mention checking optional hashes and length from the `snapshot` metadata?

For top-level targets we do say in 4.1.   *"The hashes (if any), and  version number of the new targets metadata file MUST match the trusted snapshot metadata."* 
I think the "*(if any)*" is slightly misplaced, because the hashes it refers are computed by the client. It would be more accurate to say: *"The hashes and version number of the new targets metadata file MUST match the hashes (if any) and version number listed in the trusted snapshot"*. Let me push a commit that clarifies this.

Regarding length, similar to timestamp-snapshot in 3., we say in 4.1. that the client should download targets.json up to the length specified in snapshot.json (if specified). Do you think we need another length match check in 4.1.?

Regarding delegated targets metadata, we currently don't provide details about verification. #86 adds those details and adopts the [*"(if any)"* wording](https://github.com/theupdateframework/specification/pull/86/files#diff-94b2d37a08c587082c1ea1aaead6f973R1293-R1294) used in 4.1. If we accept my suggestion for 4.1. above, we should adopt this for delegated targets metadata in  #86.

Similarly, if we add a length match check for top-level targets metadata, we should adopt this for delegated in #86 too.

Reply:
Clarified optionality of hashes (checks) in 0bdb99c. Let me know if we should add another optional length check...

Reply:
I think a second length check isn't necessary as the length is already used when downloading.

Reply:
> When we check the targets role or its delegations, do we mention checking optional hashes and length from the snapshot metadata?
FYI, @trishankatdatadog: As promised above, I just pushed a commit to #86 that notes that delegated targets metadata hashes might be missing in snapshot metadata.